### ISSUE TITLE
test_pt_ls failing on master (RHEL6)

### ISSUE THREAD

**Issue by jdetter** [dyninst/dyninst#157]
The `findMain` analysis part of this issue has been fixed, however there is a new issue causing the test to fail. We are now failing during instrumentation:

```
    Instrumenting function __GI___wcstol_l...
        Attempting to instrument function entry....
            Allocating an integer in mutatee....
                Locating BPatch_type representation for int type.... done.
                Instructing mutatee to call malloc().... done.
            done.
            Inserting information into summary report.... done.
            Retrieving list of instrumentable points from function.... done.
            Inserting counter-increment snippet into function entry.... done.
            Inserting trace instrumentation to function....
                Inserting trace instrumentation to specific point.... done.
            done.
        done.
    done.
    Instrumenting function ____wcstoul_l_internal...
        Attempting to instrument function entry....
            Allocating an integer in mutatee....
                Locating BPatch_type representation for int type.... done.
                Instructing mutatee to call malloc().... done.
            done.
            Inserting information into summary report.... done.
            Retrieving list of instrumentable points from function.... done.
            Inserting counter-increment snippet into function entry.... done.
            Inserting trace instrumentation to function....
                Inserting trace instrumentation to specific point....
                    Reading from trace pipe.... done.
                    Mutatee exited normally and returned 0
                    Reading from trace pipe.... done.
                    Mutatee exited via signal 11
Monitor received SIGCHLD
waitpid returned ff00 (255) status for pid 10294
Dyninst mutator exited normally and returned 255.
waitpid returned ff00 (255) status for pid -1
[ Done processing /bin/ls ] ---------------------------------------------------
Analysis complete.
```

@wrwilliams  or @cuviper any insight here on what could be going wrong? I am only seeing this on RHEL6 so this could be another proccontrol issue.

**Comment by wrwilliams** [comment]
Try running parseThat with -S (mutator directly available) and see if you get any more information? Mutatee segfault when reading the trace pipe suggests that there's something screwy in the RTlib/mutator combination with respect to IPC, but that's not certain.

**Comment by jdetter** [comment]
We don't get much except a message that says the mutatee exits successfully but then somehow segfaults:

```
...
01031 __acl_from_xattr Attempting to instrument function.
01141 "" Attempting to instrument function entry.
02c51 "" Allocating an integer in mutatee.
01b61 "" Locating BPatch_type representation for int type.
01b64 "" Success.
01c61 "" Instructing mutatee to call malloc().
01c64 "" Success.
02c54 "" Success.
03671 "" Inserting information into summary report.
03674 "" Success.
01f51 "" Retrieving list of instrumentable points from function.
01f54 "" Success.
02151 "" Inserting counter-increment snippet into function entry.
02154 "" Success.
02851 "" Inserting trace instrumentation to function.
02961 "" Inserting trace instrumentation to specific point.
02964 "" Success.
02854 "" Success.
01144 "" Success.
01034 "" Success.
01031 perm_copy_fd Attempting to instrument function.
01141 "" Attempting to instrument function entry.
02c51 "" Allocating an integer in mutatee.
01b61 "" Locating BPatch_type representation for int type.
01b64 "" Success.
01c61 "" Instructing mutatee to call malloc().
01c64 "" Success.
02c54 "" Success.
03671 "" Inserting information into summary report.
03674 "" Success.
01f51 "" Retrieving list of instrumentable points from function.
01f54 "" Success.
02151 "" Inserting counter-increment snippet into function entry.
02154 "" Success.
02851 "" Inserting trace instrumentation to function.
02961 "" Inserting trace instrumentation to specific point.
02a71 "" Reading from trace pipe.
02a74 "" Success.
03220 0 Mutatee exited normally.
02a71 "" Reading from trace pipe.
02a74 "" Success.
03320 11 Mutatee terminated via signal.
```

**Comment by wrwilliams** [comment]
Okay, then, try single-function instrumentation for the functions where we're crashing post-instrumentation. Is this showing up when rewriting as well, or only on a live process?

**Comment by cuviper** [comment]
I'm currently investigating a RHEL6 SIGSEGV coming from other tests -- see [rhbz 1367225](https://bugzilla.redhat.com/show_bug.cgi?id=1367225).

Perhaps the underlying cause is the same.  Does `test4_4` also SIGSEGV for you?

**Comment by jdetter** [comment]
@cuviper it looks like my `test4_4` is passing, even when running with `./test_driver -test test4_4 -v++ -allopt -allmode -allcompilers`. However, my dyninst RT lib is named `libdyninstAPI_RT.so`.

@wrwilliams it's only the create tests:

```
detter@pepperjack:testsuite$ ./test_driver -test test_pt_ls -all
TEST                       COMP   OPT  ABI MODE     THREAD  LINK    PIC     RESULT
test_pt_ls                        none 32  create   NA      dynamic nonPIC  FAILED (Running Test)
test_pt_ls                        none 64  create   NA      dynamic nonPIC  FAILED (Running Test)
test_pt_ls                        none 32  attach   NA      dynamic nonPIC  SKIPPED
test_pt_ls                        none 64  attach   NA      dynamic nonPIC  SKIPPED
test_pt_ls                        none 32  rewriter NA      dynamic nonPIC  PASSED
test_pt_ls                        none 64  rewriter NA      dynamic nonPIC  PASSED
```

**Comment by wrwilliams** [comment]
Yeah, that suggests proccontrol and/or improperly handling the instrumentation of a process mid-execution. Can you get any kind of consistent results for which functions do/don't cause the segfault when they're instrumented?

**Comment by jdetter** [comment]
It looks like it's the same function every time, however when I only instrument `____wcstoul_l_internal`, parseThat doesn't crash.

**Comment by wrwilliams** [comment]
I think runHunt works (or can be made to work) on dynamic mode as well as rewriter mode; give that a try?

**Comment by jdetter** [comment]
I got different output but it looks like it is still crashing when we instrument `____wcstoul_l_internal`, which I can't find inside of `/bin/ls` or `/lib64/libc.so.6`

```
01031 __GI___wcstol_l Attempting to instrument function.
01141 "" Attempting to instrument function entry.
02c51 "" Allocating an integer in mutatee.
01b61 "" Locating BPatch_type representation for int type.
01b64 "" Success.
01c61 "" Instructing mutatee to call malloc().
01c64 "" Success.
02c54 "" Success.
03671 "" Inserting information into summary report.
03674 "" Success.
01f51 "" Retrieving list of instrumentable points from function.
01f54 "" Success.
02151 "" Inserting counter-increment snippet into function entry.
02154 "" Success.
02851 "" Inserting trace instrumentation to function.
02961 "" Inserting trace instrumentation to specific point.
02964 "" Success.
02854 "" Success.
01144 "" Success.
01034 "" Success.
01031 ____wcstoul_l_internal Attempting to instrument function.
01141 "" Attempting to instrument function entry.
02c51 "" Allocating an integer in mutatee.
01b61 "" Locating BPatch_type representation for int type.
01b64 "" Success.
01c61 "" Instructing mutatee to call malloc().
01c64 "" Success.
02c54 "" Success.
03671 "" Inserting information into summary report.
03674 "" Success.
01f51 "" Retrieving list of instrumentable points from function.
01f54 "" Success.
02151 "" Inserting counter-increment snippet into function entry.
02154 "" Success.
02851 "" Inserting trace instrumentation to function.
02961 "" Inserting trace instrumentation to specific point.
02a71 "" Reading from trace pipe.
02a74 "" Success.
03220 0 Mutatee exited normally.
02a71 "" Reading from trace pipe.
02a74 "" Success.
03320 11 Mutatee terminated via signal.
```

**Comment by jdetter** [comment]
When running directly on `libc.so.6`, we have a failure in `processCreate`. I ran with `DYNINST_DEBUG_PROCCONTROL` and here is what I got:

https://gist.github.com/jdetter/1eed4e03394a84e2c544ff3bdbc148bd

**Comment by wrwilliams** [comment]
@jdetter any news on this one/did you see my comments on the gist?

**Comment by wrwilliams** [comment]
@mxz297 Do you have a bit of time to take a look at the CFG we're generating here and see if there's an obvious (and easy to find) chunk of missing code that would make instrumenting/relocating this function unsafe? (Related: is this fixed by the jump table fixes you've been working on?)

**Comment by mxz297** [comment]
@wrwilliams I can take a look at this tomorrow as I will be on flights later today. Could you also provide a little more details in terms of which functions should I look at?

**Comment by wrwilliams** [comment]
____wcstoul_l_internal in libc, per the comments above. I'll see if I can get quick and dirty answers today and update you; wasn't sure when your travel day was.

**Comment by wrwilliams** [comment]
...okay, so my email earlier today was looking at the wrong function, and we're having trouble with the complex internal one, not the wrapper.

**Comment by jdetter** [comment]
PR open for this issue: #286

**Comment by wrwilliams** [comment]
Fixed with merge of #286.
